Assets/Scripts/Bomb.cs

```diff
@@ -0,0 +1,39 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class Bomb : Enemy {
+    // These are only initial values for behaviour construction
+    private const float _MAX_SPEED = 4;
+    private const float _ACCELERATION_RATE = 1;
+
+    private Rigidbody2D _player;
+
+    // Bomb does nothing while idle, approaches player otherwise
+    protected override void Start() {
+        this._myRigidbody = this.GetComponent<Rigidbody2D>();
+        this._player = GameObject.FindGameObjectWithTag("Player").GetComponent<Rigidbody2D>();
+
+        // Uncomment this to check InvalidEnemyStateException checking
+        //this._idleBehaviour = new ApproachTarget(this, _player, Bomb._MAX_SPEED, Bomb._ACCELERATION_RATE, ENEMY_STATE.idle);
+
+        this._idleBehaviour = new DoNothing(this, ENEMY_STATE.idle);
+        this._awareBehaviour = new ApproachTarget(this, this._player, Bomb._MAX_SPEED, Bomb._ACCELERATION_RATE, ENEMY_STATE.aware);
+        this._engagedBehaviour = new ApproachTarget(this, this._player, Bomb._MAX_SPEED, Bomb._ACCELERATION_RATE, ENEMY_STATE.engaged);
+    }
+
+    // Bomb aware of player within 100 units, engaged within 50.
+    protected override void ChangeState() {
+        float dist_to_player = (this._myRigidbody.position - this._player.position).magnitude;
+        if(dist_to_player < 50) {
+            this._state = ENEMY_STATE.engaged;
+        } else if(dist_to_player < 100) {
+            this._state = ENEMY_STATE.aware;
+        } else {
+            this._state = ENEMY_STATE.idle;
+        }
+
+        // Use the following line of code to check that _state changes
+        // Debug.Log("My distance to the player is " + dist_to_player + " so my current enemy state is " + this._state);
```

Comment:
I wonder if we could instead use some global debug pre-processor macro, e.g.
```
#if DEBUG
Debug.Log("DEBUG");
#endif
```

that way we can enable/disable it on the project-level instead of individually via comments

Reply:
damn it, deleted my response here by mistake.

what if instead we used a public boolean instance variable that defaults to false, but can be changed in the Inspector? That way, if we know we want something in debug mode at runtime beforehand, we can change it before running, and also if we realize mid-runtime that we need some debug output logged due to unexpected behaviour, then we can change it at runtime.

If you want, I could also make a script tied to the Enemies gameobject (the parent of all the individual enemy objects) that has its own debug instance variable that can be checked. Then if that parent debug variable is true when an enemy performs its start method, I'll have the enemy switch its own debug variable to true so that all enemies in the project will be in debug mode with one variable change.

Reply:
I want to avoid using runtime checks since they are expensive. That branching can be predicted easily usually, but I imagine w/ how Unity calls on many different hooks each tick/frame, the branch likely won't reside on cache long enough to be optimized by the CPU, making it costly. Meanwhile macros are determined at compile-time, so during run-time they incur no cost.

Reply:
Is the runtime cost not worth the benefit of the ability to reactively enable debug mode mid-runtime? If we isolate debug code to `Start()` and `FixedUpdate()`, then we're doing a maximum of one `if`-statement per entity on their first frame and a maximum of 1 additional `if`-statement per entity per physics tick.

I'd say that the benefit likely outweighs the cost. One `if`-statement is likely not going to be the bottleneck of a `FixedUpdate` for an `Enemy`, and being able to enable debug output at runtime is a super-useful feature. If you still think the cost is significant enough to be worth considering, than I can run testing like we did with switches and dictionaries to see just how much more computationally efficient the macros are, but I don't think that one `if`-statement per `FixedUpdate` is going to be the straw that breaks the CPU's back.

Reply:
We discussed this in discord and decided to go for pre-processor macros due to the scalability issues of lots of conditionals being normalized across the project for debugging purposes due to problems like branch-misses causing suboptimal performance and memory management.

Reply:
Alright, I added a pre-processor macro solution in place of the stupid comment solution I was using before. Suggestion implemented.